bx_py_utils/dbperf/query_recorder.py

```diff
@@ -0,0 +1,150 @@
+        queries_duplicated = defaultdict(lambda: defaultdict(int))
+        num_queries_similar = 0
+        num_queries_duplicated = 0
+
+        # put every query into similar/duplicated dicts to aggregate their execution count
+        for alias, query in self._queries:
+            aliases.add(alias)
+            queries_similar[alias][self._similar_key(query)] += 1
+            queries_duplicated[alias][self._duplicate_key(query)] += 1
+
+        # now that the aggregation is done we must strip all queries in the
+        # aggregation dicts that were executed only once.
+        for agg_dict in (queries_similar, queries_duplicated):
+            for db, queries in agg_dict.items():
+                del_keys = []
+                for query, count in queries.items():
+                    if count == 1:
+                        del_keys.append(query)
+                # map(lambda key: queries.pop(key), del_keys)
+                for key in del_keys:
+                    del queries[key]
+
+        # for convenience, make a total for each aggregation across all databases and queries
+        for db in queries_similar:
+            num_queries_similar += len(queries_similar[db])
+        for db in queries_duplicated:
+            num_queries_duplicated += len(queries_duplicated[db])
+
+        results['queries_similar'] = queries_similar
+        results['queries_duplicated'] = queries_duplicated
+        results['num_queries_similar'] = num_queries_similar
+        results['num_queries_duplicated'] = num_queries_duplicated
+
+    def dump(self, aggregate_queries=True):
+        results = {
+            'queries': self._queries,
+            'databases': self._databases,
+            'sql_time': self._sql_time,
+            'num_queries': self._num_queries,
+        }
+
+        if aggregate_queries:
+            self._aggregate(results)
+
+        return results
+
+
+class SQLQueryRecorder:
+    """
+    A context manager that allows recording SQL queries executed during its lifetime.
+    Results of the recording can be read from the recorder in various formats after
+    recording has stopped.
+
+    In environments where not all databases in settings.DATABASES are available (e.g. unittests)
+    you should pass the databases you want to observe to the constructor.
+
+    Example usage:
+
+        with SQLQueryRecorder() as rec:
+            func_that_makes_queries()
+        print(rec.results(aggregate_results=True))
+
+    """
+    running = None
+
+    def __init__(self, databases: Optional[List[str]] = None):
+        self.logger = Logger()
+
+        if databases:
+            self.databases = [db for db in connections.all() if db.alias in databases]
+        else:
+            self.databases = connections.all()
+
+    def __enter__(self):
+        for connection in self.databases:
+            # wrap Django's default cursors in RecordingCursorWrapper
+            if not hasattr(connection, '_recording_cursor'):
```

Comment:
this can be removed, isn't it?

or better replace with a `assert not hasattr(connection, '_recording_cursor')`